wasmtime-runtime/src/jit_dump.rs

```diff
@@ -0,0 +1,650 @@
+        let filename = format!("./jit-{}.dump", process::id());
+        let mut jitdump_file;
+        unsafe {
+            let filename_c = CString::new(filename).unwrap();
+            let fd = open(
+                filename_c.as_ptr(),
+                libc::O_CREAT | libc::O_TRUNC | libc::O_RDWR,
+                0666,
+            );
+            let pgsz = sysconf(libc::_SC_PAGESIZE) as usize;
+            mmap(
+                0 as *mut c_void,
+                pgsz,
+                libc::PROT_EXEC | libc::PROT_READ,
+                libc::MAP_PRIVATE,
+                fd,
+                0,
+            );
+            jitdump_file = File::from_raw_fd(fd);
+        }
+        JitDumpAgent::write_file_header(&mut jitdump_file);
+
+        Ok(Self {
+            jitdump_file: jitdump_file,
+            code_index: 0,
+            dump_funcs: true,
+        })
+    }
+
+    fn get_time_stamp(timestamp: &mut u64) -> c_int {
+        unsafe {
+            let mut ts = mem::MaybeUninit::zeroed().assume_init();
+            if clock_gettime(libc::CLOCK_MONOTONIC, &mut ts) != 0 {
+                println!("Error getting timestamp");
+            }
+            // TODO: What does it mean for either sec or nsec to be negative?
+            *timestamp = (ts.tv_sec * 1000000000 + ts.tv_nsec) as u64;
+        }
+        return 0;
+    }
+
+    // Get the ELF machine architecture.
+    fn get_e_machine() -> u32 {
+        match target_lexicon::HOST.architecture {
+            Architecture::X86_64 => elf::header::EM_X86_64 as u32,
+            Architecture::I686 => elf::header::EM_386 as u32,
+            Architecture::Arm => elf::header::EM_ARM as u32,
+            Architecture::Armv4t => elf::header::EM_ARM as u32,
+            Architecture::Armv5te => elf::header::EM_ARM as u32,
+            Architecture::Armv7 => elf::header::EM_ARM as u32,
+            Architecture::Armv7s => elf::header::EM_ARM as u32,
+            Architecture::Aarch64 => elf::header::EM_AARCH64  as u32,
+            _ => unimplemented!("unrecognized architecture"),
+        }
+    }
+
+    fn write_file_header(file: &mut File) -> () {
+        let mut header: FileHeader = Default::default();
+        let mut timestamp: u64 = 0;
+        JitDumpAgent::get_time_stamp(&mut timestamp);
+        header.timestamp = timestamp;
+
+        // Get arch value
+        let e_machine = JitDumpAgent::get_e_machine();
+        if e_machine != elf::header::EM_NONE as u32 {
+            header.e_machine = e_machine;
+        }
+
+        // Remaning header values
+        if cfg!(target_endian = "little") {
+            header.magic = 0x4A695444
+        } else {
+            header.magic = 0x4454694a
+        }
+        header.version = 1;
+        header.size = mem::size_of::<FileHeader>() as u32;
+        header.pad1 = 0;
+        header.pid = process::id();
+        header.flags = 0;
+        bincode::serialize_into(file, &header).unwrap();
```

Comment:
I dont think bincode is a good choice. The docs of the fileformat are not defined in terms of bincode, so it is just an accident that this works. Maybe use scroll instead. scroll_derive is useful when working with it. https://github.com/m4b/scroll/blob/master/scroll_derive/README.md

Reply:
Ok, thanks for the suggestion. Definitely want to use a tool that is better suited for this type of lowering. So for example instead of:

`bincode::serialize_into(file, &header)`

instead use:

```
let mut header_bytes = vec![0; ::std::mem::size_of::<FileHeader>()];
header_bytes.pwrite_with(header, 0, NATIVE).unwrap();
file.write_all(&header_bytes);
```
Is there a more preferred way scroll is used?

Reply:
I think you can use:

```rust
file.iowrite_with(header, NATIVE).unwrap();
```

iowrite_with uses write_all internally:

https://github.com/m4b/scroll/blob/eeeeb8d1cc467400479f0d88aed9ef970268bf2f/src/lesser.rs#L162